spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Graph.scala

```diff
@@ -175,26 +186,110 @@ class Graph[T: ClassTag](val inputs : Seq[ModuleNode[T]],
       "A graph container should not be changed after it is constructed")
   }
 
+  // input node for backward graph
+  val backInputs = new Array[ModuleNode[T]](inputs.length)
+
   // Add a dummy output node, to get an one end graph. So the nodes that are not dependent by
   // the outputs will be excluded
   private val dummyOutput = new ModuleNode[T](new Dummy[T]())
+  // dummy output of backward graph
+  private var dummyOutputGrad: ModuleNode[T] = null
   outputs.foreach(_ -> dummyOutput)
 
   /**
    * Computing backgraph
    */
   val backGraph = dummyOutput.graph(reverse = true)
+  var gradGraph: DirectedGraph[AbstractModule[Activity, Tensor[T], T]] = null
```

Comment:
backGraph is used for forward, gradGraph is used for backward. Since the backward graph may be different from forward graph, we need a new graph to get backward executions